### PR TITLE
Generalize Buffer.BlockCopy optimization for either src or dst

### PR DESCRIPTION
**Pull request by jamesqo**
Don't have that much experience in native programming, but here it goes...

I noticed from #3118 that an optimization was added in the case where both `src` and `dst` were byte arrays, but not when only one of them was. Therefore, the optimization would not apply to code like this:

```cs
var bytes = new byte[4] { 1, 2, 3, 4 };
var shorts = new shorts[2];
Buffer.BlockCopy(bytes, 0, shorts, 0, bytes.Length);
```

This PR changes that so the optimization runs if either the `src` or `dst` arrays is a byte array, otherwise falling back to the default implementation as usual.

cc @jkotas

### CODE REVIEW

**Comment by jkotas** [comment]
I have never seen Buffer.BlockCopy usage pattern like this. Do you have any examples of code where this helps?

Checking for a special case - that pretty much never happens - is making the code slower for everybody else. It is not net loss - not making any real world workload faster.

**Comment by jamesqo** [comment]
@jkotas Sure, it's one of the top answers on `Buffer.BlockCopy` on StackOverflow: http://stackoverflow.com/a/3783189/4077294

Besides (correct me if I'm wrong) I don't think this is adding any overhead to the cases where this optimization doesn't apply.

File: src/vm/comutilnative.cpp [hunk 1]
```diff
@@ -1424,28 +1424,40 @@ FCIMPL5(VOID, Buffer::BlockCopy, ArrayBase *src, int srcOffset, ArrayBase *dst,
...
-
-        if (src != dst) {
-            const CorElementType dstET = dst->GetArrayElementType();
-            if (!CorTypeInfo::IsPrimitiveType_NoThrow(dstET))
-                FCThrowArgumentVoid(W("dest"), W("Arg_MustBePrimArray"));
-            dstLen = dst->GetNumComponents() * dst->GetComponentSize();
-        }
+    }
+    
+    // Optimization: If copying to/fro the same array, then
```
Comment (jkotas) [inline src/vm/comutilnative.cpp]: Typo: fro -> from

**Comment by jkotas** [comment]
I see. LGTM. Thanks!

**Pull request merged**